File: solutions/695. Max Area of Island.cpp
class Solution {
    int H,W;
    vector<vector<int>> grid;
    int dfs(int row, int col) {
        if((row < 0 || row >= H )|| (col < 0 || col >= W) || grid[row][col]==0) {
            return 0;
        }
        grid[row][col]=0;
​
        int count = 1;
        count += dfs(row+1,col) + dfs(row-1,col) + dfs(row,col+1) + dfs(row,col-1);
        return count;
    };
​
public:
    int maxAreaOfIsland(vector<vector<int>>& matrix) {
        grid = matrix;
        H = matrix.size(), W = matrix[0].size();
        int maxarea = 0;
        for (int row = 0; row < H; row++) {
            for (int col = 0; col < W; col++) {
                if (grid[row][col]) {
                    maxarea = max(maxarea, dfs(row, col));
                }
            }
        }
        return maxarea;
    }
};

File: solutions/695. Max Area of Island.java
class Solution {
    private class Pair {
        int x;
        int y;
​
        Pair(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
​
    private int[] dx = new int[] { 1, -1, 0, 0 };
    private int[] dy = new int[] { 0, 0, 1, -1 };
    private int maxarea = 0;
​
    public void area(int[][] grid, Pair start, boolean[][] visited){
        int n = grid.length;
        int m = grid[0].length;
        Queue<Pair> queue = new LinkedList<>();
        if(visited[start.x][start.y] == true) return;
        queue.add(start);
        int curarea = 1;
        while (queue.isEmpty() == false) {
            Pair p = queue.remove();
            visited[p.x][p.y] = true;
            for (int k = 0; k < 4; k++) {
                int nx = p.x + dx[k];
                int ny = p.y + dy[k];
​
                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (grid[nx][ny] == 1 && visited[nx][ny] == false) {
                        curarea++;
                        visited[nx][ny] = true;
                        queue.add(new Pair(nx, ny));
                    }
                }
            }
            
        }
        // System.out.println( start.x + " " + start.y + " " + curarea);
        this.maxarea = Math.max(curarea, this.maxarea);
    }
    
    public int maxAreaOfIsland(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        boolean[][] visited = new boolean[n][m];
        
​
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    Pair start = new Pair(i, j);
                    area(grid, start, visited);
                }
            }
        }
​
        
​
        return this.maxarea;
    }
}
